src/vsr/superblock_free_set_encoded.zig

```diff
@@ -0,0 +1,431 @@
+        }
+
+        pub fn open(
+            set: *Self,
+            grid: *Grid,
+            reference: FreeSetReference,
+            callback: *const fn (set: *Self) void,
+        ) void {
+            set.grid = grid;
+            assert(!set.grid.?.superblock.free_set.opened);
+
+            assert(set.callback == .none);
+            defer assert(set.callback == .open);
+
+            assert(set.size == 0);
+            assert(set.size_transferred == 0);
+            assert(set.block_count == 0);
+            assert(set.block_index == 0);
+
+            set.size = reference.size;
+            set.size_transferred = 0;
+            set.callback = .{ .open = callback };
+
+            set.block_count = stdx.div_ceil(set.size, chunk_size_max);
+
+            if (set.size == 0) {
+                assert(reference.head_address == 0);
+                set.block_index = 0;
+                set.grid.?.on_next_tick(open_next_tick, &set.next_tick);
+            } else {
+                assert(reference.head_address != 0);
+                // Start from the last block, as the linked list arranges data in the reverse order.
+                set.block_index = set.block_count - 1;
+                set.open_read_next(reference.head_address, reference.head_checksum);
+            }
+        }
+
+        fn open_next_tick(next_tick: *Grid.NextTick) void {
+            const set = @fieldParentPtr(Self, "next_tick", next_tick);
+            assert(set.callback == .open);
+            assert(set.block_count == 0);
+            assert(set.size == 0);
+            set.open_done();
+        }
+
+        fn open_read_next(set: *Self, address: u64, checksum: u128) void {
+            assert(set.callback == .open);
+            assert(set.block_index < set.block_count);
+            assert(address != 0);
+            assert((set.size_transferred == 0) == (set.block_index == set.block_count - 1));
+
+            set.block_addresses[set.block_index] = address;
+            set.block_checksums[set.block_index] = checksum;
+
+            set.grid.?.read_block(
+                .{ .from_local_or_global_storage = open_read_next_callback },
+                &set.read,
+                address,
+                checksum,
+                .{ .cache_read = true, .cache_write = false },
+            );
+        }
+
+        fn open_read_next_callback(read: *Grid.Read, block: Grid.BlockPtrConst) void {
+            const set = @fieldParentPtr(Self, "read", read);
+            assert(set.callback == .open);
+            assert(set.block_index < set.block_count);
+
+            const encoded_words = schema.FreeSetNode.encoded_words(block);
+            const chunk = Chunk.for_block(.{
+                .block_index = set.block_index,
+                .block_count = set.block_count,
+                .free_set_size = set.size,
+            });
+
+            stdx.copy_disjoint(
+                .exact,
+                u8,
+                set.buffer[chunk.start..chunk.end],
```

Comment:
Yeah, I am unsure here. Originally it spelled something equivalent to `[chunk.offset..chunk.size]`, which is buggy, unless there’s only a single block. 

I was quite scared by that bug (as it happened before I figured the fragmentate trick for fuzzing), and tried to design an API which is less prone to indexing errors. The result is the Chunk API, with the idea that we compute “denormalized” form with free fields so that the call-sites could be most natural.

Reply:
I don't feel strongly about it; `Chunk.for_block()` is a simple enough function that I think it is fine either way.

Reply:
Added a comment with the "why"!